### PR TITLE
Optimize `ZEnvironment#prune` for single service tags

### PR DESCRIPTION
**Pull request by kyri-petrou**
It's fairly common to prune the `ZEnvironment` using a single tag, especially when composing ZLayers. This PR optimizes this path by avoiding unnecessary allocations that are necessary when pruning an environment for multiple tags as well as some other optimizations that are applicable to both single-service and multi-service pruning.

Also fixed a bug where using `prune[Scope]` on an environment that doesn't contain a Scope wouldn't throw an error.

Added benchmark shows ~200-300% throughput increase.

### CODE REVIEW

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala [hunk 1]
```diff
@@ -192,38 +203,31 @@ private[zio] object UpdateOrderLinkedMap {
...
+    private[this] var entries: List[(K, V)] = Nil
 
     def addOne(elem: (K, V)): UpdateOrderLinkedMap.Builder[K, V] = {
-      if (aliased ne null) {
-        aliased = aliased.updated(elem._1, elem._2)
-      } else {
-        // Place them in reverse order, we'll reverse them back during `result()`
-        entries = elem :: entries
-      }
+      entries = elem :: entries
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: Should we store keys/values separately to avoid the tuples? This class doesn't seem to extend any built in builder type that requires tuples?

edit: I see that we are using `self.map.iterator` which will produce tuples anyways...

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala [hunk 2]
```diff
@@ -192,38 +203,31 @@ private[zio] object UpdateOrderLinkedMap {
   def newBuilder[K, V]: UpdateOrderLinkedMap.Builder[K, V] = new UpdateOrderLinkedMap.Builder[K, V]
 
   final class Builder[K, V] { self =>
-    private[this] var entries: List[(K, V)]               = Nil
-    private[this] var aliased: UpdateOrderLinkedMap[K, V] = _
+    private[this] var entries: List[(K, V)] = Nil
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: Would it be more efficient to use a GrowableArray or ListBuffer?

File: core/shared/src/main/scala/zio/ZEnvironment.scala [hunk 3]
```diff
@@ -131,21 +131,56 @@ final class ZEnvironment[+R] private (
    * contained within it.
    */
   def prune[R1 >: R](implicit tagged: EnvironmentTag[R1]): ZEnvironment[R1] = {
-    val tag = taggedTagType(tagged)
 
-    // Mutable set lookups are much faster. It also iterates faster. We're better off just allocating here
-    // Why are immutable set lookups so slow???
-    val set = new mutable.HashSet ++= taggedGetServices(tag)
+    def throwError(tags: collection.Set[LightTypeTag]) =
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/ZEnvironment.scala:135]: Does moving this method out reduce the byte code size by much?

File: core/shared/src/main/scala/zio/ZEnvironment.scala [hunk 4]
```diff
@@ -131,21 +131,56 @@ final class ZEnvironment[+R] private (
...
+      throw new Error(
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/ZEnvironment.scala:136]: pre-existing: should we have an error type for defects?

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala [hunk 5]
```diff
@@ -192,38 +203,31 @@ private[zio] object UpdateOrderLinkedMap {
...
-        var reversed  = List.empty[(K, V)]
-        var remaining = entries
-        val set       = mutable.HashSet.empty[K]
-        while (remaining ne Nil) {
-          val head = remaining.head
-          if (set.add(head._1)) reversed = head :: reversed
-          remaining = remaining.tail
-        }
-        aliased = fromUnsafe(reversed.iterator)
+    def result(): UpdateOrderLinkedMap[K, V] =
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: To confirm, we only ever expect `result()` to be called once? which the previous `aliased` optimization was not assuming?

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala [hunk 6]
```diff
@@ -87,7 +87,9 @@ private[zio] final class UpdateOrderLinkedMap[K, +V](
     if (self.fields.size - size > 10000) fromUnsafe(iterator0)
     else self
 
-  def iterator: Iterator[(K, V)] = iteratorLz.iterator
+  def iterator: Iterator[(K, V)] =
+    if (size == 1) underlying.iterator.map { case (k, v) => (k, v.value) }
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: Is it possible to directly call `apply` based on the `fields(0)`? or is allocating the iterator cheaper?

File: core/shared/src/main/scala/zio/ZEnvironment.scala [hunk 7]
```diff
@@ -131,21 +131,56 @@ final class ZEnvironment[+R] private (
...
+      val it0 = self.map.iterator
+      while (it0.hasNext) {
+        val next    = it0.next()
+        val leftTag = next._1
+
+        if (taggedIsSubtype(leftTag, tag)) {
+          builder addOne next
+        }
+      }
+      val needsScope = isScopeTag(tag)
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/ZEnvironment.scala]: How likely is it that we are pruning to just scope? Would it be best to fast path this separately? It seems like the common case is either we need a newMap or we need scope - not boith

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala (hunk 1 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: Yeah we need to use tuples. Even if we stored them separately, we'd be adding to two lists which will create an object per entry anyways

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala (hunk 2 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: It depends how strict we want to be. If we can guarantee that the keys being added to the builder are already unique, then it'll be more optimized to use a `ListBuffer`.

Taking a look at our current usages of it, I think we can optimize. Let me take a look if I can do it in a safe-ish way.

File: core/shared/src/main/scala/zio/ZEnvironment.scala (hunk 3 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZEnvironment.scala:135]: By "out" you mean not having it nested within the `prune` method? The scala compiler does that anyways because nested methods are not supported at the JVM level, so they become private methods on the parent class instead.

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala (hunk 5 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: Yes that's right 👍 When I initially wrote the implementation I followed how other builders are commonly written in the Scala std lib, but tbh that was a big overkill as we're not planning on making these classes public

File: core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala (hunk 6 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala]: We could call `head` and then `Iterator.single` though; I think that would save 1 function invocation / allocation. Let me make the change.

File: core/shared/src/main/scala/zio/ZEnvironment.scala (hunk 7 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZEnvironment.scala]: It shouldn't be that common but I don't see a reason not to optimize that path. I'll add it

File: core/shared/src/main/scala/zio/ZEnvironment.scala (hunk 4 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZEnvironment.scala:136]: Maybe, but let's leave it for the time being, I don't think it's that much of an issue since it should never happen

File: core/shared/src/main/scala/zio/ZEnvironment.scala [hunk 8]
```diff
@@ -131,65 +131,123 @@ final class ZEnvironment[+R] private (
...
 
         if (set.contains(leftTag)) {
           // Exact match, no need to loop
           found.add(leftTag)
           builder addOne next
         } else {
           // Need to check whether it's a subtype
-          var loop = true
-          val it1  = set.iterator
-          while (it1.hasNext && loop) {
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZEnvironment.scala:158]: @hearnadam I realised that this was being called _A LOT_, especially on large environments. It's much better to create a list and iterate using a `var` instead. Even if that means we're allocating a new List, it's much better than creating a new iterator for all entries in the environment

File: core/shared/src/main/scala/zio/ZEnvironment.scala (hunk 3 above)
Comment (hearnadam) [reply core/shared/src/main/scala/zio/ZEnvironment.scala:135]: Ah you're right

File: core/shared/src/main/scala/zio/ZEnvironment.scala [hunk 9]
```diff
@@ -131,65 +131,123 @@ final class ZEnvironment[+R] private (
...
+      }
+
+    def pruneMany(iSet: Set[LightTypeTag]): ZEnvironment[R1] = {
+      // Mutable set lookups are much faster. It also iterates faster. We're better off just allocating here
+      // Why are immutable set lookups so slow???
+      val set        = new mutable.HashSet ++= iSet
+      val tagsAsList = iSet.toList // Lists iterate much faster than sets
+      val builder    = UpdateOrderLinkedMap.newBuilder[LightTypeTag, Any]
+      val found      = new mutable.HashSet[LightTypeTag]
+      val nil        = Nil
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/ZEnvironment.scala:185]: Does this avoid a reference load or something?

File: core/shared/src/main/scala/zio/ZEnvironment.scala [hunk 10]
```diff
@@ -131,65 +131,123 @@ final class ZEnvironment[+R] private (
...
 
       if (set.size > found.size) {
-        val missing = set -- found
+        val missing = set --= found
 
         // We need to check whether one of the services we added is a subtype of the missing service
         val newTags = newMap.keySet
         missing.foreach { tag =>
           if (newTags.exists(taggedIsSubtype(_, tag))) missing.remove(tag)
         }
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/ZEnvironment.scala:233]: Can we use the `fields` here as a semi-strict keyset to avoid allocating a keySet?

**Review by hearnadam — APPROVED** [review]

File: core/shared/src/main/scala/zio/ZEnvironment.scala (hunk 9 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZEnvironment.scala:185]: It's kind of a trick I saw being used in `jsoniter-scala`. I'm not sure how useful it actually is though; I'd imagine the JVM will likely inline the object reference instead of loading it on each iteration.

File: core/shared/src/main/scala/zio/ZEnvironment.scala (hunk 10 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZEnvironment.scala:233]: I'm not sure I understand your recommendation. Which `fields` are you referring to?

Also, I think we shouldn't put too much effort into optimizing this branch. It's very unlikely to reach this part here as it's there only to accommodate for this edge case: https://github.com/zio/zio/blob/5c58c4981f8ad11ebd7fcecfe0af944385c02763/core-tests/shared/src/test/scala/zio/ZEnvironmentSpec.scala#L51-L60
